### ISSUE TITLE
Support for GE Smart Whole Home Water Filter

### ISSUE THREAD

**Issue by tumtumsback** [simbaja/ha_gehome#21]
Model = GXWH70M

https://www.geapplianceparts.com/store/parts/spec/GXWH70M

I have this installed in my house, and I have it added to my SmartHQ app. If you could integrate that into this custom component, that would be amazing. Thanks!

**Comment by simbaja** [comment]
We actually should have support for that in the dev branch.  We'd love some assistance in testing it out if you have a chance.

**Comment by tumtumsback** [comment]
So I deleted the integration, copied the link from the dev url, pasted that into HACS, added that GE Home integration in HACS, referenced that in the regular integrations area, put in my SmartHQ creds, but it's not populating the GE Smart Whole Home Water Filter as an entity within HA. Is there something else I need to be doing? Thanks!

**Comment by simbaja** [comment]
Did any errors show up in the log?  We haven't tested the dev branch much yet (trying to get all the new functionality in, then will harden).  If not, is there any chance you can go to the sdk (simbaja\gehome) and grab a debug log?  Another possibility is to set the debug level in HA to debug, the entries should show up from HA, but will be mixed in with other HA messages so harder to see.

**Comment by tumtumsback** [comment]
I think what might be happening is the Dev branch isn't popping into HACS properly. When I put the dev url "https://github.com/simbaja/ha_components/tree/dev" into HACS, it just references "https://github.com/simbaja/ha_components" ... anything I need to do specifically to make sure I'm using the dev branch in HACS?

**Comment by simbaja** [comment]
You might have to install manually by copying the files into the custom components folder.  I'm not sure HACS can deal with development branches, but not sure.

**Comment by tumtumsback** [comment]
![image](https://user-images.githubusercontent.com/58952365/125477866-43983320-db37-4fb9-ad22-4e38362feaf4.png)

**Comment by tumtumsback** [comment]
Sweet -- got it in there. One of the entities shows as not available. What would be really cool is if there would be a way that I could toggle between Valve State for Filter/Bypass/Off. That way, I can create an automation in tandem with a zigbee leak sensor I have nearby that would basically put the valve into "Off" mode if the leak sensor went from Dry --> Wet

**Comment by warrenrees** [comment]
What version is your home assistant? If it's not 2021.7.1 you should upgrade and it should let you control the filter.

**Comment by tumtumsback** [comment]
I'm on 2021.7.2 -- what would be a good way to control the filter valve state? I'm wondering from both an automation perspective, and a dashboard/card perspective. Thank you so much btw!

**Comment by simbaja** [comment]
The missing entity is the one that allows control.  Are there any errors logged for showing why it's unavailable?

**Comment by tumtumsback** [comment]
`Logger: homeassistant.components.select
Source: components/select/__init__.py:90
Integration: Select (documentation, issues)
First occurred: 11:12:34 AM (2 occurrences)
Last logged: 11:12:34 AM

Error adding entities for domain select with platform ge_home
Error while setting up ge_home platform for select
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 383, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 588, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 666, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 419, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 453, in _async_write_ha_state
    state = self._stringify_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 425, in _stringify_state
    state = self.state
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 78, in state
    if self.current_option is None or self.current_option not in self.options:
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 90, in current_option
    return self._attr_current_option
AttributeError: 'GeErdFilterPositionSelect' object has no attribute '_attr_current_option'`

**Comment by simbaja** [comment]
Looks like I missed a required override for the select entity.  Try now and let me know if it shows up.

**Comment by tumtumsback** [comment]
Logger: homeassistant.components.select
Source: components/select/__init__.py:90
Integration: Select (documentation, issues)
First occurred: 2:59:43 PM (2 occurrences)
Last logged: 2:59:43 PM

Error adding entities for domain select with platform ge_home
Error while setting up ge_home platform for select
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 383, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 588, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 666, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 419, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 453, in _async_write_ha_state
    state = self._stringify_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 425, in _stringify_state
    state = self.state
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 78, in state
    if self.current_option is None or self.current_option not in self.options:
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 90, in current_option
    return self._attr_current_option
AttributeError: 'GeErdFilterPositionSelect' object has no attribute '_attr_current_option'

**Comment by simbaja** [comment]
Try one more time, I think I missed property decorators so my overrides weren't taking effect.

**Comment by tumtumsback** [comment]
Logger: homeassistant.components.select
Source: components/select/__init__.py:90
Integration: Select (documentation, issues)
First occurred: 3:13:21 PM (2 occurrences)
Last logged: 3:13:21 PM

Error adding entities for domain select with platform ge_home
Error while setting up ge_home platform for select
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 383, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 588, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 666, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 419, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 453, in _async_write_ha_state
    state = self._stringify_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 425, in _stringify_state
    state = self.state
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 78, in state
    if self.current_option is None or self.current_option not in self.options:
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 90, in current_option
    return self._attr_current_option
AttributeError: 'GeErdFilterPositionSelect' object has no attribute '_attr_current_option'

**Comment by simbaja** [comment]
That's odd.  That attribute should no longer matter from what I'm seeing.  Did you fully restart HA after updating with the latest copy from dev?  I might have to look a little deeper after work if so because I'm not seeing the issue at the moment.

**Comment by tumtumsback** [comment]
Hang on let me try again

**Comment by tumtumsback** [comment]
Logger: homeassistant.components.select
Source: components/select/__init__.py:78
Integration: Select (documentation, issues)
First occurred: 4:15:16 PM (2 occurrences)
Last logged: 4:15:16 PM

Error adding entities for domain select with platform ge_home
Error while setting up ge_home platform for select
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 383, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 588, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 666, in add_to_platform_finish
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 419, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 453, in _async_write_ha_state
    state = self._stringify_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 425, in _stringify_state
    state = self.state
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 78, in state
    if self.current_option is None or self.current_option not in self.options:
TypeError: argument of type 'method' is not iterable

**Comment by simbaja** [comment]
I found another property decorator I missed.  Hopefully that's the last one...

**Comment by tumtumsback** [comment]
SO CLOSE!

![image](https://user-images.githubusercontent.com/58952365/125664851-c89a2a5a-e3f6-4444-876f-0f3b3ad390f2.png)

**Comment by tumtumsback** [comment]
Just won't let me choose the position -- it's greyed out ...

**Comment by simbaja** [comment]
Sorry, long week, hopefully should work now.  Thanks for your patience on this.

**Comment by tumtumsback** [comment]
All good -- I know how it goes :)

Do I need to redownload the dev file(s) and replace them in my custom_components?

**Comment by simbaja** [comment]
Yep, that should do it.

**Comment by tumtumsback** [comment]
OK I'll try right now -- super excited

**Comment by tumtumsback** [comment]
Ok so now the entity isn’t greyed out but when I choose bypass or off or whatever nothing actually happens to the filter itself and/or in the SmartHQ app

**Comment by simbaja** [comment]
Interesting, the water filter support was based on a commit from @bendavis.  I've modified it a bit to make the select logic a little more generic (more to do there), but otherwise it's basically the same as what was committed (see https://github.com/simbaja/ha_components/pull/16#issue-679192511).

It's possible I messed up support during some of my updates, but it's also possible you have a different model than @bendavis that works differently.  Is it possible to run the raw sdk (simbaja\gehome) to get us some debug logs to help diagnose?

@bendavis, can you run the current dev version and see if it's working for you?  Hopefully I didn't screw it up, but certainly possible :).

**Comment by simbaja** [comment]
@tumtumsback, not sure if this is possible for you, but to narrow things down a little, you can also try @bendavis's original version to see if I messed it up somewhere along the line.  You'd need to pull down commit 459fc7f5cb8c21ab6b3d54670aeb4ac816a5f0f0 instead of the current version.  Not sure how you're doing the clones, but this may help to get this version: https://stackoverflow.com/questions/7832770/how-to-get-certain-commit-from-github-project

**Comment by simbaja** [comment]
> Ok so now the entity isn’t greyed out but when I choose bypass or off or whatever nothing actually happens to the filter itself and/or in the SmartHQ app

Also, forgot to ask, any errors in the log?

**Comment by simbaja** [comment]
I made a few corrections to the "dev" branch.  I've got another user that has confirmed everything is essentially working except for the position select setting.  When you have a moment, can you confirm what is working for you?

**Comment by bendavis** [comment]
hi @simbaja, @tumtumsback, sorry I've been preoccupied with home life. The corrections I had are mostly to import statements since HACS was loading from a different location than the way I was expecting. The select stuff is new - you need to be a on a recent branch of HA, and even then I wasn't getting the UI to render correctly, but the service calls were working for me.

**Comment by bendavis** [comment]
@simbaja - I'm trying to catch back up on the mods - I had created a filter_life_remaining and filter_rate_sensor to get the units of measurements correct. Now they are just generic sensors again? I had problems being able to plot them (not an number) before. I also see that @tumtumsback is getting Unknown in those values? 

Here's my running instance:
![image](https://user-images.githubusercontent.com/532255/126345177-7213bfcf-992c-43a7-90c4-0a04a8230c4c.png)

**Comment by simbaja** [comment]
Hmm... if you're using the latest dev branch (from last night), all that should be fixed up.  There were a few issues in the way that I converted back to generic sensors.  We should be able to use generic sensors and just add erd code classes to handle the correct uom.  That way as we expand functionality, we can re-use the uom stuff and have to develop less customized sensors.  I've been working with @rgabrialson11 as well to figure this functionality out and it seems to be working for him, but perhaps not?  See: https://github.com/simbaja/ha_components/issues/18#issue-940323688 for more discussions on this.  Towards the end are his debug logs for the filter.

**Comment by simbaja** [comment]
> hi @simbaja, @tumtumsback, sorry I've been preoccupied with home life. The corrections I had are mostly to import statements since HACS was loading from a different location than the way I was expecting. The select stuff is new - you need to be a on a recent branch of HA, and even then I wasn't getting the UI to render correctly, but the service calls were working for me.

No worries :)  We all get busy (I was preoccupied with work for a week or so too).  

I think that the UI is rendering correctly for the select entity now, it just isn't working to set the position for some reason.  It's almost as if it's not actually sending the ERD set command (see the other thread for logs).  I can see the debug message that it's trying to set it, but looks like nothing ever makes it to the appliance.  I posted an example to set it to off over there to see if something was up with the SDK, but I suspect it's something with the select.  If you have a chance to look through and debug the set, that would be appreciated.

**Comment by simbaja** [comment]
@bendavis , also, forgot to mention, if the way I converted to generic sensors doesn't allow plotting, can you take a look at the ErdPropertySensor?  That should also probably work.  I did see that you adjusted the state a bit in your commit, which should be handled by the ErdSensor, but not 100% sure if it'll all work the way I hope.

**Comment by simbaja** [comment]
@tumtumsback, @bendavis: Looks like @rgabrielson11 and I have gotten everything working, including the position setting from the HA select entity.  I just added a couple protections on the movement (i.e. can only move when in ready state), hopefully I didn't break anything when I did that.  Can you both give the dev version a shot and let me know how it looks to you?

**Comment by tumtumsback** [comment]
You guys are AWESOME! I am now able to switch between Filtered/Bypass/Off via HA... THANK YOU SO MUCH

**Comment by simbaja** [comment]
@tumtumsback: Can you confirm that it disables the select properly as it's transitioning between states and does not allow you to set it again?  Can you also confirm that the numeric values are being interpreted correctly by HA?  Things like flow rate, daily usage, and life remaining should all appear as graphs in the logbook.

**Comment by tumtumsback** [comment]
It does not disable the select during transition

I am a bad candidate for testing value correctness -- in my SmartHQ app, it stopped showing me the GPM, like months and months ago -- nothing to do with HA. I have no idea why

I personally just wanted this custom integration to work for an automation I wanted, where I could have the valve on the GE Water Filter turn to "off" if certain zigbee water sensors in my house detected moisture.

Any chance you've seen these filters stop reporting GPM stats, even just on the SmartHQ side of things?

**Comment by simbaja** [comment]
> It does not disable the select during transition

For the select, does it show "Transition" during the transition?  Or does it still have the same values like normal?  Might have to force an update or something if it doesn't switch...

> Any chance you've seen these filters stop reporting GPM stats, even just on the SmartHQ side of things?

I actually don't have a water filter.  From what @rgabrielson11 said, it looks like the time of registration affects how the stats are produced.  My thought is that you may want to remove/re-add the device in SmartHQ and see if it helps.

**Comment by tumtumsback** [comment]
Just removed and re-added ... still not luck yet -- maybe it takes some time 🤷

**Comment by tumtumsback** [comment]
![image](https://user-images.githubusercontent.com/58952365/126726666-883c2862-3ba4-4435-afd0-b2f91d90eafb.png)

**Comment by tumtumsback** [comment]
Yes, I see "Transition" for Filter Mode, and "Driving" for Valve State. The "Filter Position" dropdown does not grey-out during the transition, though. Once the transition is done, it goes back to "Ready" though.

**Comment by rgabrielson11** [comment]
I just tested it and confirm @tumtumsback is correct.  Select does not grey out, however selecting a different mode while in transition does not have any impact, the second selection is ignored and state is the first selection.

**Comment by rgabrielson11** [comment]
> It does not disable the select during transition
> 
> I am a bad candidate for testing value correctness -- in my SmartHQ app, it stopped showing me the GPM, like months and months ago -- nothing to do with HA. I have no idea why
> 
> I personally just wanted this custom integration to work for an automation I wanted, where I could have the valve on the GE Water Filter turn to "off" if certain zigbee water sensors in my house detected moisture.
> 
> Any chance you've seen these filters stop reporting GPM stats, even just on the SmartHQ side of things?

I would go to https://accounts.brillion.geappliances.com/ and deactivate everything under 'my apps'.  Then, next time you log in using the phone app it will ask you to reauthorize.  From what I have seen, this will reset all of your usage counters.  Try to do this as close as possible to midnight however, as it starts the 24 hour usage cycle based on this activation time.  Unless you don't care about if the 24 hour usage resets at whatever time you activate.

**Comment by simbaja** [comment]
Assuming that everything is working in latest versions, please feel free to re-open if something isn't working.
